### PR TITLE
[marker] store data and clean on file deletion

### PR DESCRIPTION
**Pull request by svenefftinge**
A marker-service will remove markers when a file gets deleted.
A marker-service will persist marker information between sessions.

### CODE REVIEW

File: packages/markers/src/browser/marker-manager.ts [hunk 1]
```diff
@@ -117,6 +126,69 @@ export abstract class MarkerManager<D extends object> {
     protected readonly uri2MarkerCollection = new Map<string, MarkerCollection<D>>();
     protected readonly onDidChangeMarkersEmitter = new Emitter<void>();
 
+    initialized: Promise<void>;
```
Comment (akosyakov) [inline packages/markers/src/browser/marker-manager.ts]: readonly?

File: packages/core/src/browser/frontend-application.ts [hunk 2]
```diff
@@ -95,7 +104,17 @@ export class FrontendApplication {
...
-            contribution.onStart(this);
+            if (contribution.onStart) {
+                contribution.onStart(this);
+            }
+        }
+
+        window.onunload = () => {
+            for (const contribution of this.contributions.getContributions()) {
+                if (contribution.onStop) {
+                    contribution.onStop(this);
```
Comment (akosyakov) [inline packages/core/src/browser/frontend-application.ts]: Should we do try/catch here? that once contribution does not break others

File: packages/core/src/browser/frontend-application.ts [hunk 3]
```diff
@@ -9,16 +9,26 @@ import { inject, injectable, named } from 'inversify';
...
  */
 export const FrontendApplicationContribution = Symbol("FrontendApplicationContribution");
 export interface FrontendApplicationContribution {
+
+    /**
+     * Called when the application is started.
+     */
+    onStart?(app: FrontendApplication): void;
+
+
```
Comment (hexa00) [inline packages/core/src/browser/frontend-application.ts]: empty line

File: packages/core/src/browser/frontend-application.ts [hunk 4]
```diff
@@ -9,16 +9,26 @@ import { inject, injectable, named } from 'inversify';
...
     /**
-     * Callback
+     * Called when an application is stopped or unloaded.
+     *
+     * Note that this is implemented using `window.unload` which doesn't allow any asynchronous code anymore. I.e. this is the last tick.
      */
-    onStart(app: FrontendApplication): void;
+    onStop?(app: FrontendApplication): void
```
Comment (hexa00) [inline packages/core/src/browser/frontend-application.ts]: missing semi-column

File: packages/core/src/browser/shell-layout-restorer.ts [hunk 5]
```diff
@@ -15,34 +15,34 @@ import { ILogger } from '../common/logger';
...
+
     constructor(
         @inject(WidgetManager) protected widgetManager: WidgetManager,
         @inject(ILogger) protected logger: ILogger,
         @inject(StorageService) protected storageService: StorageService) { }
 
     onStart(app: FrontendApplication): void {
-        const storageKey = 'layout';
-        this.storageService.getData(storageKey, undefined).then((serializedLayoutData: string | undefined) => {
+        this.storageService.getData(this.storageKey, undefined).then((serializedLayoutData: string | undefined) => {
```
Comment (hexa00) [inline packages/core/src/browser/shell-layout-restorer.ts]: You don't need to specify the undefined default... but I don't mind

File: packages/markers/src/browser/marker-manager.ts [hunk 6]
```diff
@@ -117,6 +126,69 @@ export abstract class MarkerManager<D extends object> {
     protected readonly uri2MarkerCollection = new Map<string, MarkerCollection<D>>();
     protected readonly onDidChangeMarkersEmitter = new Emitter<void>();
```
Comment (hexa00) [inline packages/markers/src/browser/marker-manager.ts]: empty line

**Comment by svenefftinge** [comment]
I need this change for the next PR. Can I merge?

File: packages/core/src/browser/frontend-application.ts (hunk 3 above)
Comment (akosyakov) [reply packages/core/src/browser/frontend-application.ts]: https://palantir.github.io/tslint/rules/no-consecutive-blank-lines/ can catch such cases

Comment (hexa00) [reply packages/core/src/browser/frontend-application.ts]: Nice I'll add that thx

Comment (svenefftinge) [reply packages/core/src/browser/frontend-application.ts]: Would also be good if our autoformatting would clean that already.

Comment (hexa00) [reply packages/core/src/browser/frontend-application.ts]: Yes let's fix that as part of #413

**Pull request merged**